python/lsst/ts/observatory/control/mock/base_group_mock.py

```diff
@@ -0,0 +1,254 @@
+        self._output_only = []
+
+        for i, component in enumerate(self._components):
+            name, index = salobj.name_to_name_index(component)
+            rname = component.lower().replace(":", "_")
+            self._component_names.append(rname)
+            _controllers[rname] = salobj.Controller(name=name, index=index)
+            if component in output_only:
+                self._output_only.append(rname)
+
+        self._component_names = tuple(self._component_names)
+        self._output_only = tuple(self._output_only)
+
+        self.controllers = types.SimpleNamespace(**_controllers)
+
+        self.setting_versions = {}
+
+        self.settings_to_apply = {}
+
+        for comp in self.component_names:
+            if comp not in self._output_only:
+                getattr(
+                    self.controllers, comp
+                ).cmd_start.callback = self.get_start_callback(comp)
+                getattr(
+                    self.controllers, comp
+                ).cmd_enable.callback = self.get_enable_callback(comp)
+                getattr(
+                    self.controllers, comp
+                ).cmd_disable.callback = self.get_disable_callback(comp)
+                getattr(
+                    self.controllers, comp
+                ).cmd_standby.callback = self.get_standby_callback(comp)
+
+        for comp in self.output_only:
+            for cmd in getattr(self.controllers, comp).salinfo.command_names:
+                getattr(
+                    getattr(self.controllers, comp), f"cmd_{cmd}"
+                ).callback = self.generic_raise_callback
+
+        self.start_task = asyncio.create_task(self.start_task_publish())
+
+        self.run_telemetry_loop = False
+
+        self.task_list = []
+
+    async def start_task_publish(self):
+
+        self.run_telemetry_loop = True
+
+        if self.start_task.done():
+            raise RuntimeError("Start task already completed.")
+
+        await asyncio.gather(
+            *[
+                getattr(self.controllers, name).start_task
+                for name in self._component_names
+            ]
+        )
+
+        for comp in self.component_names:
+            if comp not in self.output_only:
+                getattr(self.controllers, comp).evt_summaryState.set_put(
+                    summaryState=salobj.State.STANDBY
+                )
+                self.setting_versions[comp] = f"test_{comp}"
+                getattr(self.controllers, comp).evt_settingVersions.set_put(
+                    recommendedSettingsVersion=f"{self.setting_versions[comp]},"
+                )
+
+    async def generic_callback(self, data):
+        await asyncio.sleep(HEARTBEAT_INTERVAL)
+
+    async def generic_raise_callback(self, data):
+        """ A generic callback function that will raise an exception if called.
+        """
+        raise RuntimeError("This command should not be called.")
+
+    def get_start_callback(self, comp):
```

Comment:
Please consider rewriting these as the actual callback methods and using `functools.partial` to send the `comp` argument when you specify the method as the callback function.

I _think_ it would make the code significantly easier to read.

For example:
```
# Define the method
async def start_callback(self, data, comp):
    ss = getattr(self.controllers, comp).evt_summaryState
    ...
...
# Use the method:
getattr(
    self.controllers, comp
).cmd_start.callback = functools.partial(self.get_start_callback, comp=comp))
```